File: LeetCode/permutation.js
/**
 * https://leetcode.com/problems/permutations/description/
 * Difficulty: medium
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    if(nums.length <= 1){
        return [nums];
    }
    var permutations = [];
    for(var i = 0; i < nums.length; i++){
        var tmpNum = nums[i];
        if(nums.indexOf(tmpNum ) !== i){
            continue;
        }
        var remainingNums = nums.slice(0,i).concat(nums.slice(i+1,nums.length));
        for(var tmpPermutation of permute(remainingNums)){
            console.log(tmpPermutation);
            permutations.push([tmpNum].concat(tmpPermutation));
        }
    }
    return permutations;

};

File: LeetCode/permutation.rb
# https://leetcode.com/problems/permutations/description/
# difficulty: medium
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
   if nums.length <= 1
    return [nums]
   end
   i = 0
   ret = []
   for i in 0..nums.length
    tmpNum = nums[i]
    if nums.index(tmpNum).to_i<i
      next
    end
    remainingNums = nums.slice(0,i).concat(nums.slice(i+1,nums.length))
    permute(remainingNums).each{|tmpPermutation|
      ret.push([tmpNum].concat(tmpPermutation))
    }
  end
  return ret
end

myArr = permute([1,2,3])
myArr.each{|group|
  puts group
}